File: code/TemplateStudioForUWP/Templates/_comp/P/IdentityCommon/Param_ProjectName/Services/UserDataService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Param_RootNamespace.Core.Helpers;
using Param_RootNamespace.Core.Models;
using Param_RootNamespace.Core.Services;
using Param_RootNamespace.Helpers;
using Param_RootNamespace.ViewModels;
using Windows.Storage;

namespace Param_RootNamespace.Services
{
    public class UserDataService : IUserDataService
    {
        private const string _userSettingsKey = "IdentityUser";
        private UserViewModel _user;
        private IIdentityService _identityService;
        private IMicrosoftGraphService _microsoftGraphService;

        public event EventHandler<UserViewModel> UserDataUpdated;

        public UserDataService(IIdentityService identityService, IMicrosoftGraphService microsoftGraphService)
        {
            _identityService = identityService;
            _microsoftGraphService = microsoftGraphService;
        }

        public void Initialize()
        {
            _identityService.LoggedIn += OnLoggedIn;
            _identityService.LoggedOut += OnLoggedOut;
        }

        public async Task<UserViewModel> GetUserAsync()
        {
            if (_user == null)
            {
                _user = await GetUserFromCacheAsync();
                if (_user == null)
                {
                    _user = GetDefaultUserData();
                }
            }

            return _user;
        }

        private async void OnLoggedIn(object sender, EventArgs e)
        {
            _user = await GetUserFromGraphApiAsync();
            UserDataUpdated?.Invoke(this, _user);
        }

        private async void OnLoggedOut(object sender, EventArgs e)
        {
            _user = null;
            await ApplicationData.Current.LocalFolder.SaveAsync<User>(_userSettingsKey, null);
        }

        private async Task<UserViewModel> GetUserFromCacheAsync()
        {
            var cacheData = await ApplicationData.Current.LocalFolder.ReadAsync<User>(_userSettingsKey);
            return await GetUserViewModelFromData(cacheData);
        }

        private async Task<UserViewModel> GetUserFromGraphApiAsync()
        {
            var accessToken = await _identityService.GetAccessTokenForGraphAsync();
            if (string.IsNullOrEmpty(accessToken))
            {
                return null;
            }

            var userData = await _microsoftGraphService.GetUserInfoAsync(accessToken);
            if (userData != null)
            {
                userData.Photo = await _microsoftGraphService.GetUserPhoto(accessToken);
                await ApplicationData.Current.LocalFolder.SaveAsync(_userSettingsKey, userData);
            }

            return await GetUserViewModelFromData(userData);
        }

        private async Task<UserViewModel> GetUserViewModelFromData(User userData)
        {
            if (userData == null)
            {
                return null;
            }

            var userPhoto = string.IsNullOrEmpty(userData.Photo)
                ? ImageHelper.ImageFromAssetsFile("DefaultIcon.png")
                : await ImageHelper.ImageFromStringAsync(userData.Photo);

            return new UserViewModel()
            {
                Name = userData.DisplayName,
                UserPrincipalName = userData.UserPrincipalName,
                Photo = userPhoto
            };
        }

        private UserViewModel GetDefaultUserData()
        {
            return new UserViewModel()
            {
                Name = _identityService.GetAccountUserName(),
                Photo = ImageHelper.ImageFromAssetsFile("DefaultIcon.png")
            };
        }
    }
}

File: code/TemplateStudioForUWP/Templates/_comp/MT/IdentityCommon._VB/Services/UserDataService.vb
Imports System.Collections.Generic
Imports System.Threading.Tasks
Imports Param_RootNamespace.Core.Helpers
Imports Param_RootNamespace.Core.Models
Imports Param_RootNamespace.Core.Services
Imports Param_RootNamespace.Helpers
Imports Param_RootNamespace.ViewModels
Imports Windows.Storage

Namespace Services
    Public Class UserDataService
        Private Const _userSettingsKey As String = "IdentityUser"
        Private _user As UserViewModel

        Private ReadOnly Property IdentityService As IdentityService
            Get
                Return Singleton(Of IdentityService).Instance
            End Get
        End Property

        Private ReadOnly Property MicrosoftGraphService As MicrosoftGraphService
            Get
                Return Singleton(Of MicrosoftGraphService).Instance
            End Get
        End Property

        Public Event UserDataUpdated As EventHandler(Of UserViewModel)

        Public Sub New()
        End Sub

        Public Sub Initialize()
            AddHandler IdentityService.LoggedIn, AddressOf OnLoggedIn
            AddHandler IdentityService.LoggedOut, AddressOf OnLoggedOut
        End Sub

        Public Async Function GetUserAsync() As Task(Of UserViewModel)
            If _user Is Nothing Then
                _user = Await GetUserFromCacheAsync()

                If _user Is Nothing Then
                    _user = GetDefaultUserData()
                End If
            End If

            Return _user
        End Function

        Private Async Sub OnLoggedIn(sender As Object, e As EventArgs)
            _user = Await GetUserFromGraphApiAsync()
            RaiseEvent UserDataUpdated(Me, _user)
        End Sub

        Private Async Sub OnLoggedOut(sender As Object, e As EventArgs)
            _user = Nothing
            Await ApplicationData.Current.LocalFolder.SaveAsync(Of User)(_userSettingsKey, Nothing)
        End Sub

        Private Async Function GetUserFromCacheAsync() As Task(Of UserViewModel)
            Dim cacheData = Await ApplicationData.Current.LocalFolder.ReadAsync(Of User)(_userSettingsKey)
            Return Await GetUserViewModelFromData(cacheData)
        End Function

        Private Async Function GetUserFromGraphApiAsync() As Task(Of UserViewModel)
            Dim accessToken = Await IdentityService.GetAccessTokenForGraphAsync()

            If String.IsNullOrEmpty(accessToken) Then
                Return Nothing
            End If

            Dim userData = Await MicrosoftGraphService.GetUserInfoAsync(accessToken)

            If userData IsNot Nothing Then
                userData.Photo = Await MicrosoftGraphService.GetUserPhoto(accessToken)
                Await ApplicationData.Current.LocalFolder.SaveAsync(_userSettingsKey, userData)
            End If

            Return Await GetUserViewModelFromData(userData)
        End Function

        Private Async Function GetUserViewModelFromData(userData As User) As Task(Of UserViewModel)
            If userData Is Nothing Then
                Return Nothing
            End If

            Dim userPhoto = If(String.IsNullOrEmpty(userData.Photo), ImageHelper.ImageFromAssetsFile("DefaultIcon.png"), Await ImageHelper.ImageFromStringAsync(userData.Photo))
            Return New UserViewModel() With {
                .Name = userData.DisplayName,
                .UserPrincipalName = userData.UserPrincipalName,
                .Photo = userPhoto
            }
        End Function

        Private Function GetDefaultUserData() As UserViewModel
            Return New UserViewModel() With {
                .Name = IdentityService.GetAccountUserName(),
                .Photo = ImageHelper.ImageFromAssetsFile("DefaultIcon.png")
            }
        End Function
    End Class
End Namespace
